src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs

```diff
@@ -0,0 +1,139 @@
+                    {
+                        return null;
+                    }
+                }
+                else
+                {
+                    // The versions we have in our cache (if any) do not match the ones provided by the client (if any).
+                    // We need to calculate new results.
+                    cheapVersion = await cache.ComputeCheapVersionAsync(state, cancellationToken).ConfigureAwait(false);
+                    expensiveVersion = await cache.ComputeExpensiveVersionAsync(state, cancellationToken).ConfigureAwait(false);
+                }
+
+                // Compute the new result for the request.
+                var data = await cache.ComputeDataAsync(state, cancellationToken).ConfigureAwait(false);
+                var dataHash = GetComputedDataHash(data);
+
+                string newResultId;
+                if (_lastResult is not null && _lastResult?.resultId == previousPullResult?.PreviousResultId && _lastResult?.hashedData == dataHash)
+                {
+                    // The new data we've computed is exactly the same as the data we computed last time even though the versions have changed.
+                    // Instead of reserializing everything, we can return the same result id back to the client.
+
+                    // Ensure we store the updated versions we calculated with the old resultId.
+                    _lastResult = (_lastResult.Value.resultId, cheapVersion, expensiveVersion, dataHash);
+                    return null;
+                }
+                else
+                {
+                    // Keep track of the results we reported here so that we can short-circuit producing results for
+                    // the same state of the world in the future.  Use a custom result-id per type (doc requests or workspace
+                    // requests) so that clients of one don't errantly call into the other.
+                    //
+                    // For example, a client getting document diagnostics should not ask for workspace diagnostics with the result-ids it got for
+                    // doc-diagnostics.  The two systems are different and cannot share results, or do things like report
+                    // what changed between each other.
+                    //
+                    // Note that we can safely update the map before computation as any cancellation or exception
+                    // during computation means that the client will never recieve this resultId and so cannot ask us for it.
+                    newResultId = $"{uniqueKey}:{cache.IncrementResultId()}";
+                    _lastResult = (newResultId, cheapVersion, expensiveVersion, dataHash);
+                    return (newResultId, data);
+                }
+            }
+        }
+
+        private static int GetComputedDataHash(ImmutableArray<TComputedData> data)
+        {
+            var hashes = data.SelectAsArray(d => d.GetHashCode()).Sort();
```

Comment:
needs docs.  first, why are you sorting?

second.  are these types HashCode safe?  Not convinced they are.  i have some thoughts here.  overall idea is sound, but if we don't make this part bulletproof we may not send updates when we need to.